connect/api/src/main/java/org/apache/kafka/connect/data/Values.java

```diff
@@ -0,0 +1,1168 @@
+        if (value instanceof Long) {
+            return Schema.INT64_SCHEMA;
+        }
+        if (value instanceof Float) {
+            return Schema.FLOAT32_SCHEMA;
+        }
+        if (value instanceof Double) {
+            return Schema.FLOAT64_SCHEMA;
+        }
+        if (value instanceof byte[]) {
+            return Schema.BYTES_SCHEMA;
+        }
+        if (value instanceof List) {
+            List<?> list = (List<?>) value;
+            if (list.isEmpty()) {
+                return null;
+            }
+            SchemaDetector detector = new SchemaDetector();
+            for (Object element : list) {
+                if (!detector.canDetect(element)) {
+                    return null;
+                }
+            }
+            return SchemaBuilder.array(detector.schema()).build();
+        }
+        if (value instanceof Map) {
+            Map<?, ?> map = (Map<?, ?>) value;
+            if (map.isEmpty()) {
+                return null;
+            }
+            SchemaDetector keyDetector = new SchemaDetector();
+            SchemaDetector valueDetector = new SchemaDetector();
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                if (!keyDetector.canDetect(entry.getKey()) || !valueDetector.canDetect(entry.getValue())) {
+                    return null;
+                }
+            }
+            return SchemaBuilder.map(keyDetector.schema(), valueDetector.schema()).build();
+        }
+        if (value instanceof Struct) {
+            return ((Struct) value).schema();
+        }
+        return null;
+    }
+
+
+    /**
+     * Parse the specified string representation of a value into its schema and value.
+     *
+     * @param value the string form of the value
+     * @return the schema and value; never null, but whose schema and value may be null
+     * @see #convertToString
+     */
+    public static SchemaAndValue parseString(String value) {
+        if (value == null) {
+            return NULL_SCHEMA_AND_VALUE;
+        }
+        if (value.isEmpty()) {
+            return new SchemaAndValue(Schema.STRING_SCHEMA, value);
+        }
+        Parser parser = new Parser(value);
+        return parse(parser, false);
+    }
+
+    /**
+     * Convert the value to the desired type.
+     *
+     * @param toSchema   the schema for the desired type; may not be null
+     * @param fromSchema the schema for the supplied value; may be null if not known
+     * @return the converted value; never null
+     * @throws DataException if the value could not be converted to the desired type
+     */
+    protected static Object convertTo(Schema toSchema, Schema fromSchema, Object value) throws DataException {
+        if (value == null) {
+            if (toSchema.isOptional()) {
+                return null;
+            }
+            throw new DataException("Unable to convert a null value to a non-null value");
+        }
+        switch (toSchema.type()) {
+            case BYTES:
```

Comment:
Right, but that while that does end up making each method less complex, the expansion into many methods makes the class harder to really understand the potential paths. Seems like we'd just be playing games with checkstyle and not _really_ simplifying the code.

Reply:
yeah, people differ in their opinions about granularity and understandability. I think I personally align more w/ you, where just inlining stuff seems more understandable overall.

if others agree, maybe things like cyclomatic complexity can simply have increased limits instead of requiring breaking these things up more. that said, this code is definitely an extreme example -- i had to work a bit to track alignment to make sure I understood where, and at what depth, various blocks were.